File: cses/04_graph_algorithms/1750_planets-queries-i/solution.py
import sys


def main():
    data = sys.stdin.buffer.read().split()
    idx = 0
    n = int(data[idx]); idx += 1
    q = int(data[idx]); idx += 1

    # t: đích (0-based) của teleporter duy nhất tại mỗi hành tinh.
    t = [int(data[idx + i]) - 1 for i in range(n)]
    idx += n

    LOG = 30  # 2^30 > 1e9 >= k_max

    # Binary lifting: up[j][v] = đỉnh đến được sau 2^j bước từ v; up[0] = t.
    # up[j][v] == up[j-1][up[j-1][v]] == prev[prev[v]] cho mọi v, nên cả mức tính
    # bằng list(map(prev.__getitem__, prev)) -- nhanh hơn hẳn vòng lặp Python.
    up = [t]
    for _ in range(1, LOG):
        prev = up[-1]
        up.append(list(map(prev.__getitem__, prev)))

    # Đọc mọi truy vấn vào các mảng phẳng (điểm xuất phát 0-based).
    xs = [int(data[idx + 2 * i]) - 1 for i in range(q)]
    ks = [int(data[idx + 2 * i + 1]) for i in range(q)]

    # Xử lý gộp: tại mức j chỉ dịch những truy vấn có bit j của k bật, dùng up[j].
    # Batch tránh vòng while từng truy vấn và dồn phần nặng vào tầng C.
    cur = xs
    for j in range(LOG):
        upj = up[j]
        cur = [upj[c] if (kk >> j) & 1 else c for c, kk in zip(cur, ks)]

    # Cộng lại 1 để khớp định dạng 1-indexed của CSES, gom in một lần.
    sys.stdout.write("\n".join(map(str, (c + 1 for c in cur))) + ("\n" if cur else ""))


if __name__ == "__main__":
    main()

File: cses/04_graph_algorithms/1750_planets-queries-i/solution.cpp
#include <iostream>
#include <vector>

using namespace std;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n, queries;
    cin >> n >> queries;

    // Binary lifting: successor[level][v] = đỉnh đến được sau 2^level bước từ v.
    // Cần 2^LEVELS > k_max = 1e9, chọn LEVELS = 30.
    constexpr int LEVELS = 30;
    vector<vector<int>> successor(LEVELS, vector<int>(n + 1));

    // successor[0][v] = t[v]: đích của teleporter duy nhất tại mỗi hành tinh.
    for (int planet = 1; planet <= n; ++planet) {
        cin >> successor[0][planet];
    }

    // Dựng bảng theo quy nạp: đi 2^level = ghép hai lần đi 2^(level-1).
    for (int level = 1; level < LEVELS; ++level) {
        for (int planet = 1; planet <= n; ++planet) {
            successor[level][planet] =
                successor[level - 1][successor[level - 1][planet]];
        }
    }

    // Trả lời từng truy vấn: phân tích steps theo nhị phân, mỗi bit bật thì nhảy.
    while (queries--) {
        int planet, steps;
        cin >> planet >> steps;
        for (int level = 0; level < LEVELS; ++level) {
            if (steps & (1 << level)) {
                planet = successor[level][planet];
            }
        }
        cout << planet << '\n';
    }
    return 0;
}
